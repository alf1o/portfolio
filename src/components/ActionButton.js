import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import Message from '@material-ui/icons/Message';
import Work from '@material-ui/icons/Work';
import Code from '@material-ui/icons/Code';
import Mail from '@material-ui/icons/Mail';
import { Link } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';

const styles = theme => ({
  trigger: {
    position: 'absolute',
    backgroundColor: theme.palette.primary.main
  },
  actions: {
    position: 'absolute',
    transition: 'bottom, right 300ms ease-out',
    backgroundColor: '#fff'
  }
});

function getIcon(icon) {
  switch(icon) {
    case 'message':
      return <Message color="secondary" />;
    case 'mail':
      return <Mail color="primary" />
    case 'linkedin':
      return <Work color="primary" />
    case 'github':
      return <Code color="primary" />
    default:
      return <Message color="secondary" />;
  }
}

/**
  Render a button which will open a way of contanct.
*/
function ActionButton({ classes, pos, icon, onClick }) {
  return (
    <Button
      className={
        icon === 'message' ? classes.trigger : classes.actions
      }
      style={{ bottom: pos.bottom, right: pos.right }}
      variant="fab"
      onClick={onClick}
    >
      {getIcon(icon)}
    </Button>
  );
}

ActionButton.propTypes = {
  classes: PropTypes.object.isRequired,
  pos: PropTypes.object.isRequired,
  icon: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired
};

export { ActionButton };
export default withStyles(styles)(ActionButton);
