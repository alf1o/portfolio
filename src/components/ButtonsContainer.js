import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ActionButton from './ActionButton';
import { withStyles } from '@material-ui/core/styles';

const styles = {

};

/**
  Container for the action buttons.
*/
class ButtonsContainer extends Component {

  static propTypes = {
    classes: PropTypes.object.isRequired
  };

  constructor(props) {
    super(props);
    this.state = {
      btnsOpen: false
    };
    this.toggleBtns = this.toggleBtns.bind(this);
  }

  toggleBtns() {
    this.setState(prevState => ({ btnsOpen: !prevState.btnsOpen }));
  }

  render() {
    const { classes } = this.props;
    const { btnsOpen } = this.state;
    const defaultPos = { bottom: 20, right: 15 };
    return (
      <div>
        <ActionButton
          pos={
            btnsOpen
            ? { bottom: 85, right: 7 }
            : defaultPos
          }
          icon="mail"
          onClick={this.toggleBtns}
        /><ActionButton
          pos={
            btnsOpen
            ? { bottom: 73, right: 72 }
            : defaultPos
          }
          icon="linkedin"
          onClick={this.toggleBtns}
        /><ActionButton
          pos={
            btnsOpen
            ? { bottom: 9, right: 83 }
            : defaultPos
          }
          icon="github"
          onClick={this.toggleBtns}
        />
        <ActionButton
          pos={defaultPos}
          icon="message"
          onClick={this.toggleBtns}
        />
      </div>
    );
  }
}

export { ButtonsContainer };
export default withStyles(styles)(ButtonsContainer);
