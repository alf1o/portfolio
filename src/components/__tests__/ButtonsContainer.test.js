import React from 'react';
import ButtonsContainer from '../ButtonsContainer';
import { createShallow } from '@material-ui/core/test-utils';
import ActionButton from '../ActionButton';

describe('`ButtonsContainer`', () => {
  let props = {};
  let mountedBtns;
  const shallow = createShallow();
  const defaultPos = { bottom: 20, right: 15 };

  function btns() {
    if (!mountedBtns) mountedBtns = shallow(<ButtonsContainer {...props} />);
    return mountedBtns;
  }

  beforeEach(() => {
    props = {};
    mountedBtns = undefined;
  });

  it('should pass `pos`, `icon` and `onClick` to each `ActionButton`', () => {
    const actions = btns().find(ActionButton);
    actions.forEach(action => {
      expect(action.props().pos).toBeDefined();
      expect(action.props().icon).toBeDefined();
      expect(action.props().onClick).toBeDefined();
    });
  });

  it('should have a `state.btnsOpen` property', () => {
    expect(btns().state().btnsOpen).toBeDefined();
  });

  describe('`btnsOpen` is `false`', () => {
    beforeEach(() => {
      btns().setState({ btnsOpen: false });
    });

    it('should render `ActionButton`s with the `defaultPos`', () => {
      const actions = btns().find(ActionButton);
      actions.forEach(action => {
        expect(action.props().pos).toEqual(defaultPos);
      });
    });
  });

  it('should have a `toggleBtns` method', () => {
    expect(ButtonsContainer.prototype.toggleBtns).toBeDefined();
  });

  it('should pass `toggleBtns` to each `ActionButton`', () => {
    const actions = btns().find(ActionButton);
    const toggleBtns = btns().instance().toggleBtns;
    actions.forEach(action => {
      expect(action.props().onClick).toEqual(toggleBtns);
    });
  });

  describe(`toggleBtns`, () => {
    const toggleBtnsSpy = jest.spyOn(ButtonsContainer.prototype, 'toggleBtns');
    afterEach(() => {
      toggleBtnsSpy.mockClear();
    });

    it('should toggle `btnsOpen`', () => {
      expect(btns().state().btnsOpen).toBe(false);
      toggleBtnsSpy.call(btns().instance());
      expect(btns().state().btnsOpen).toBe(true);
      toggleBtnsSpy.call(btns().instance());
      expect(btns().state().btnsOpen).toBe(false);
    });
  });

  it('should match snapshot', () => {
    expect(btns()).toMatchSnapshot();
  });
});
