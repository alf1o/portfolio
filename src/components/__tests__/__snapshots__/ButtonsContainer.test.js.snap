// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`ButtonsContainer\` should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ButtonsContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <a
          href="mailto:alfioparisi.93@gmail.com"
>
          <WithStyles(ActionButton)
                    icon="mail"
                    onClick={[Function]}
                    pos={
                              Object {
                                        "bottom": 20,
                                        "right": 15,
                                      }
                    }
          />
</a>,
        <a
          href="https://uk.linkedin.com/in/alfio-parisi-2b3baa13b"
>
          <WithStyles(ActionButton)
                    icon="linkedin"
                    onClick={[Function]}
                    pos={
                              Object {
                                        "bottom": 20,
                                        "right": 15,
                                      }
                    }
          />
</a>,
        <a
          href="https://github.com/alf1o"
>
          <WithStyles(ActionButton)
                    icon="github"
                    onClick={[Function]}
                    pos={
                              Object {
                                        "bottom": 20,
                                        "right": 15,
                                      }
                    }
          />
</a>,
        <WithStyles(ActionButton)
          icon="message"
          onClick={[Function]}
          pos={
                    Object {
                              "bottom": 20,
                              "right": 15,
                            }
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <WithStyles(ActionButton)
            icon="mail"
            onClick={[Function]}
            pos={
                        Object {
                                    "bottom": 20,
                                    "right": 15,
                                  }
            }
/>,
          "href": "mailto:alfioparisi.93@gmail.com",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "icon": "mail",
            "onClick": [Function],
            "pos": Object {
              "bottom": 20,
              "right": 15,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <WithStyles(ActionButton)
            icon="linkedin"
            onClick={[Function]}
            pos={
                        Object {
                                    "bottom": 20,
                                    "right": 15,
                                  }
            }
/>,
          "href": "https://uk.linkedin.com/in/alfio-parisi-2b3baa13b",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "icon": "linkedin",
            "onClick": [Function],
            "pos": Object {
              "bottom": 20,
              "right": 15,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <WithStyles(ActionButton)
            icon="github"
            onClick={[Function]}
            pos={
                        Object {
                                    "bottom": 20,
                                    "right": 15,
                                  }
            }
/>,
          "href": "https://github.com/alf1o",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "icon": "github",
            "onClick": [Function],
            "pos": Object {
              "bottom": 20,
              "right": 15,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "a",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "icon": "message",
          "onClick": [Function],
          "pos": Object {
            "bottom": 20,
            "right": 15,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <a
            href="mailto:alfioparisi.93@gmail.com"
>
            <WithStyles(ActionButton)
                        icon="mail"
                        onClick={[Function]}
                        pos={
                                    Object {
                                                "bottom": 20,
                                                "right": 15,
                                              }
                        }
            />
</a>,
          <a
            href="https://uk.linkedin.com/in/alfio-parisi-2b3baa13b"
>
            <WithStyles(ActionButton)
                        icon="linkedin"
                        onClick={[Function]}
                        pos={
                                    Object {
                                                "bottom": 20,
                                                "right": 15,
                                              }
                        }
            />
</a>,
          <a
            href="https://github.com/alf1o"
>
            <WithStyles(ActionButton)
                        icon="github"
                        onClick={[Function]}
                        pos={
                                    Object {
                                                "bottom": 20,
                                                "right": 15,
                                              }
                        }
            />
</a>,
          <WithStyles(ActionButton)
            icon="message"
            onClick={[Function]}
            pos={
                        Object {
                                    "bottom": 20,
                                    "right": 15,
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <WithStyles(ActionButton)
              icon="mail"
              onClick={[Function]}
              pos={
                            Object {
                                          "bottom": 20,
                                          "right": 15,
                                        }
              }
/>,
            "href": "mailto:alfioparisi.93@gmail.com",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "icon": "mail",
              "onClick": [Function],
              "pos": Object {
                "bottom": 20,
                "right": 15,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <WithStyles(ActionButton)
              icon="linkedin"
              onClick={[Function]}
              pos={
                            Object {
                                          "bottom": 20,
                                          "right": 15,
                                        }
              }
/>,
            "href": "https://uk.linkedin.com/in/alfio-parisi-2b3baa13b",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "icon": "linkedin",
              "onClick": [Function],
              "pos": Object {
                "bottom": 20,
                "right": 15,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <WithStyles(ActionButton)
              icon="github"
              onClick={[Function]}
              pos={
                            Object {
                                          "bottom": 20,
                                          "right": 15,
                                        }
              }
/>,
            "href": "https://github.com/alf1o",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "icon": "github",
              "onClick": [Function],
              "pos": Object {
                "bottom": 20,
                "right": 15,
              },
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "a",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "icon": "message",
            "onClick": [Function],
            "pos": Object {
              "bottom": 20,
              "right": 15,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {},
  },
}
`;
